midi_app_controller/controller/connected_controller.py

```diff
@@ -0,0 +1,363 @@
+
+        try:
+            # Create an instance of MidiIn and MidiOut
+            midi_in = rtmidi.MidiIn()
+            midi_out = rtmidi.MidiOut()
+
+            # Listing available MIDI ports
+            available_ports_in = midi_in.get_ports()
+            available_ports_out = midi_out.get_ports()
+
+            available_ports = [
+                port for port in available_ports_in if port in available_ports_out
+            ]
+
+            controller_port = ""
+            port_index = -1
+
+            if available_ports:
+                for i, port in enumerate(available_ports):
+                    name = port.split(":")[0]
+                    if name == controller.name:
+                        controller_port = port
+                        port_index = i
+
+            if controller_port == "":
+                raise IOError("No correct MIDI ports available.")
+
+            # Creating MidiIn and MidiOut instances
+            midi_in.open_port(port_index)
+            midi_out.open_port(port_index)
+
+        except TypeError as err:
+            print(f"Type Error: {err}")
+        except SystemError as err:
+            print(f"System Error: {err}")
+        except rtmidi.InvalidPortError as err:
+            print(f"Invalid Port Error: {err}")
+        except rtmidi.InvalidUseError as err:
+            print(f"Invalid Use Error: {err}")
+
+        instance = cls(
+            controller=controller,
+            actions_handler=actions_handler,
+            midi_out=midi_out,
+            midi_in=midi_in,
+            button_ids=button_ids,
+            button_engagement={},
+            knob_ids=knob_ids,
+            knob_engagement={},
+        )
+
+        instance.init_buttons()
+        instance.init_knobs()
+
+        # Set callback for getting data from controller
+        instance.midi_in.set_callback(midi_callback, data=instance)
+
+        return instance
+
+    def __del__(self):
+        """Responsible for closing in/out ports, and safely deleting
+        python-rtmidi classes.
+        """
+        self.midi_in.close_port()
+        self.midi_in.delete()
+        self.midi_out.close_port()
+        self.midi_out.delete()
+
+    def init_buttons(self):
+        """Initializes the buttons on the controller, setting them
+        to the 'off' value. Adds button entries to 'button_engagement'
+        dictionary.
+        """
+        for id in self.button_ids:
+            self.turn_off_button_led(id)
+            self.button_engagement[id] = self.controller.button_value_off
+
+    def init_knobs(self):
+        """Initializes the knobs on the controller, setting them
+        to the minimal value. Adds knob entries to 'knob_engagement'
+        dictionary.
+        """
+        for id in self.knob_ids:
+            self.change_knob_value(id, self.controller.knob_value_min)
+            self.knob_engagement[id] = self.controller.knob_value_min
+
+    def handle_button_engagement(self, data):
```

Comment:
Can you also add type hints for parameters and return value in function declarations? Like `def handle_button_engagement(self, data: List[int]) -> None:`. This comment applies to all the functions in this file.